### ISSUE TITLE
Too much GPU memory occupied

### ISSUE THREAD

**Issue by AlexGranger-scn** [chieupham/PCAAE#1]
I am very interested in this PCA-like AutoEncoder algorithm, and I try to apply it to extract features from voice embedding. I follow the guidance in `Introduction_PCAAE.ipynb`. I change the input data into my dataset and retain the rest parts of `Introduction_PCAAE.ipynb`,  and the code could run successfuly.

Now I face a problem that while the parameter `step` increases, the required GPU memory increases rapidly. I have reduced my batch size to 2 and the memory is still not enough. As a result, I can not finish a whole training process (actually when step comes to 2, the memory is nearly full). My experiment environment is one 1080 Ti with 11GB GPU memory.  

I wonder if any operation takes so much memory and I haven't solved this problem so far.

Could you do me a favour and enlighten me how to improve the performance of the code?

**Comment by chieupham** [comment]
Hi,
Thank you for your interest ! The introduction can run successfully using one 1080 Ti GPU card. Have you tried to remove the cell concerning the standard autoencoder `AE_E`, `AE_D`  and the PCA_AE without cov_loss as `PCAAE_E_noCov` and `PCAAE_D_noCov` to save the memory ? You can separate `Introduction_PCAAE.ipynb ` into 3 parts such as the standard AE, PCA AE and PCA AE without cov_loss to test your data. Don't forget to save the weights of each model to test after training.
Could you tell me if this helps to solve your problem ?
Best regards,

**Comment by chieupham** [comment]
Sometimes, the dataset is too big, we can not use `torch.stack` to merge all data. Another way to process this data set is to create a custom Dataset before feeding to `torch.utils.data.DataLoader`. You could find some solutions here : https://pytorch.org/tutorials/beginner/basics/data_tutorial.html or the example `class FaceLandmarksDataset(Dataset)` in this link : https://pytorch.org/tutorials/recipes/recipes/custom_dataset_transforms_loader.html.
Hope this helps !

**Comment by AlexGranger-scn** [comment]
> Sometimes, the dataset is too big, we can not use `torch.stack` to merge all data. Another way to process this data set is to create a custom Dataset before feeding to `torch.utils.data.DataLoader`. You could find some solutions here : https://pytorch.org/tutorials/beginner/basics/data_tutorial.html or the example `class FaceLandmarksDataset(Dataset)` in this link : https://pytorch.org/tutorials/recipes/recipes/custom_dataset_transforms_loader.html. Hope this helps !

Thanks for your reply! 
In my code, I have already defined a cunstom Dataset to merge my data. And `Introduction_PCAAE.ipynb ` is very clear, so I only use PCAAE_E and PCAAE_D and removed rest parts of code~
Yesterday I use `print("1:{}".format(torch.cuda.memory_allocated(0)))` to watch the cuda memory and I found in `Testing part` in class `train_PCA_AE`, the cuda memory usage has increased significantly, so I add `with torch.no_grad()` before the whole testing part~ After this modification, the cuda memory usage is acceptable~ 
Now every data is a 256 dimension vector, dtype is float32, my batchsize is 1024, encoder and decoder consist of 3 linear layers and activation functions. Under these settings, the first step only takes about 10MB. And each step increases cuda memory usage by only about 10MB as well~(I think this is much more resonable...)
Thanks for your reply again :-)

**Comment by Andr1aS** [comment]
Hello,
Thanks for sharing the code for the PCA-AE algorithm.

I am interested in applying the algorithm in other datasets as well, starting from the simple ones, such as MNIST and celebA. But the training remains constant at 87% level. Do you have any clue why this might happen?

Ps: I used the provided code for the ellipses and modified it to be trained on torchvision datasets.

**Comment by chieupham** [comment]
Hello @Andr1aS,
Thank you for your interest ! I am not sure that I could figure out your problem. Have you trried to reduce parameter lambda of the covariance loss "lambda_cov" ? Maybe, we can set lambda_cov of 0.1.
In our paper, we tested on celebA and the image outputs are very burring ! So I think that we need some recent decoder to better reconstruct the output.
I hope this helps !
Chi-Hieu
